src/base/bittorrent/session.cpp

```diff
@@ -1887,8 +1889,58 @@ bool Session::addTorrent_impl(CreateTorrentParams params, const MagnetUri &magne
 
         p = magnetUri.addTorrentParams();
     }
-    else if (torrentInfo.isValid()) {
-        if (!params.restored) {
+    else {
+        if (!torrentInfo.isValid()) {
+            // We can have an invalid torrentInfo when there isn't a matching
+            // .torrent file to the .fastresume we loaded. Possibly from a
+            // failed upgrade.
+            return false;
+        }
+
+        hash = torrentInfo.hash();
+    }
+
+    // We should not add the torrent if it is already
+    // processed or is pending to add to session
+    if (m_addingTorrents.contains(hash) || m_loadedMetadata.contains(hash))
+        return false;
+
+    TorrentHandle *const torrent = m_torrents.value(hash);
+    if (torrent) {  // a duplicate torrent is added
+        if (torrent->isPrivate() || (!fromMagnetUri && torrentInfo.isPrivate()))
+            return false;
+
+        // merge trackers and web seeds
+        torrent->addTrackers(fromMagnetUri ? magnetUri.trackers() : torrentInfo.trackers());
+        torrent->addUrlSeeds(fromMagnetUri ? magnetUri.urlSeeds() : torrentInfo.urlSeeds());
+        return true;
+    }
+
+    // Record if .fastresume file contains fetched metadata
+    // In the case of Magnet metadata isn't fetched yet, we still need
+    // to setup the default parameters.
+    bool hasCompleteFastresume = false;
+
+    if (!fromMagnetUri) {
+        if (params.restored) {  // load from existing fastresume
+#if (LIBTORRENT_VERSION_NUM < 10200)
+            p.resume_data = std::vector<char> {fastresumeData.constData()
+                , (fastresumeData.constData() + fastresumeData.size())};
+            p.flags |= lt::add_torrent_params::flag_use_resume_save_path;
+
+            // we still setup the default parameters and
+            // let libtorrent handle the parameter merging
+            hasCompleteFastresume = false;
+#else
+            lt::error_code ec;
+            p = lt::read_resume_data(fastresumeData, ec);
+
+            // fastresume will only contain "qBt-*" entries when metadata
+            // hasn't been fetched
+            hasCompleteFastresume = !p.file_priorities.empty();
```

Comment:
Are you sure that "file priorities" is mandatory for fastresumedata so libtorrent produce it in any cases? Note that some fields are omitted for, e.g., completed torrents.

Reply:
Yes, looking from libtorrent doc it should always set file priorities to torrent when metadata is available.